pkg/graphql/federation/node_extractor.go

```diff
@@ -0,0 +1,194 @@
+		fieldName := r.document.FieldDefinitionNameString(fieldRef)
+
+		if added := r.addChildTypeFieldName(typeName, fieldName, childNodes); !added {
+			continue
+		}
+
+		fieldTypeName := r.getNodeName(r.document.FieldDefinitionTypeNode(fieldRef))
+		r.findChildNodesForType(fieldTypeName, childNodes)
+	}
+}
+
+func (r *nodeExtractor) addChildTypeFieldName(typeName, fieldName string, childNodes *[]plan.TypeField) bool {
+	for i := range *childNodes {
+		if (*childNodes)[i].TypeName != typeName {
+			continue
+		}
+
+		for _, field := range (*childNodes)[i].FieldNames {
+			if field == fieldName {
+				return false
+			}
+		}
+
+		(*childNodes)[i].FieldNames = append((*childNodes)[i].FieldNames, fieldName)
+		return true
+	}
+
+	*childNodes = append(*childNodes, plan.TypeField{
+		TypeName:   typeName,
+		FieldNames: []string{fieldName},
+	})
+
+	return true
+}
+
+func (r *nodeExtractor) addRootNodes(astNode ast.Node, rootNodes *[]plan.TypeField) {
+	typeName := r.getNodeName(astNode)
+	if !r.isEntity(astNode) && !r.isRootOperationTypeName(typeName) {
+		return
+	}
+
+	var fieldNames []string
+
+	fieldRefs := r.document.NodeFieldDefinitions(astNode)
+	for _, fieldRef := range fieldRefs {
+		if isExternalField(r.document, fieldRef) {
+			continue
+		}
+
+		fieldName := r.document.FieldDefinitionNameString(fieldRef)
+		fieldNames = append(fieldNames, fieldName)
+	}
+
+	if len(fieldNames) == 0 {
+		return
+	}
+
+	*rootNodes = append(*rootNodes, plan.TypeField{
+		TypeName:   typeName,
+		FieldNames: fieldNames,
+	})
+}
+
+// document.NodeNameBytes method doesnt support NodeKindObjectTypeExtension and NodeKindInterfaceTypeExtension
+func (r *nodeExtractor) getNodeName(astNode ast.Node) string {
+	var ref ast.ByteSliceReference
+
+	switch astNode.Kind {
+	case ast.NodeKindObjectTypeDefinition:
+		ref = r.document.ObjectTypeDefinitions[astNode.Ref].Name
+	case ast.NodeKindObjectTypeExtension:
+		ref = r.document.ObjectTypeExtensions[astNode.Ref].Name
+	case ast.NodeKindInterfaceTypeExtension:
+		ref = r.document.InterfaceTypeExtensions[astNode.Ref].Name
+	case ast.NodeKindInterfaceTypeDefinition:
+		ref = r.document.InterfaceTypeDefinitions[astNode.Ref].Name
+	case ast.NodeKindInputObjectTypeDefinition:
+		ref = r.document.InputObjectTypeDefinitions[astNode.Ref].Name
+	case ast.NodeKindUnionTypeDefinition:
+		ref = r.document.UnionTypeDefinitions[astNode.Ref].Name
+	case ast.NodeKindScalarTypeDefinition:
+		ref = r.document.ScalarTypeDefinitions[astNode.Ref].Name
+	case ast.NodeKindDirectiveDefinition:
+		ref = r.document.DirectiveDefinitions[astNode.Ref].Name
+	case ast.NodeKindField:
+		ref = r.document.Fields[astNode.Ref].Name
+	case ast.NodeKindDirective:
+		ref = r.document.Directives[astNode.Ref].Name
+	}
+
+	bytesName := r.document.Input.ByteSlice(ref)
+
+	return unsafebytes.BytesToString(bytesName)
+}
+
+func (r *nodeExtractor) isRootOperationTypeName(typeName string) bool {
+	rootOperationNames := map[string]struct{}{
+		"Query":        {},
```

Comment:
What if:
```
schema {
  query: MyQuery
}
```

A similar problem is solved here: https://github.com/TykTechnologies/graphql-go-tools/blob/master/pkg/asttransform/baseschema.go#L119